README.md
# SYRENE - Syndrom-based Recommendation Network

A recommendation system based on Syndrom Networks.
Written in Literate Haskell, as an internal project
when I worked on RenderX XEP.


changes.txt
# $Id: changes.txt,v 1.31 2007/05/04 09:28:28 dvd Exp $

Fri May  4 12:13:51 IDT 2007
    alternative hypotheses done; seems to work and give good information.
    Now let's describe what is done -- write papers.

Wed Apr 25 14:13:19 IDT 2007
    sydiff is implemented - a command-line utility that takes
    new and old network and a list of cases and displays cases
    for which explanations differ.

Tue Apr 24 23:28:15 IDT 2007
    I need a tool that displays a list of cases affected by differences
    between two versions of the questionnaire. The tool takes old version,
    new version and a list of cases and returns a list of cases with different
    explanations.

    It is probably worth making it a command-line utility along with a function
    in Sysh.

Sun Apr 22 16:50:20 IDT 2007
    Active dialogue done. Utility is logged to stderr (that is, to httpd-error.log).
    Next is development tools.

Sun Apr  8 03:31:38 IDT 2007
    String values are numerically bi-valued (0,1). Ana is drafted.
    Next is Acti.

Fri Apr  6 16:30:07 IDT 2007
    Exposed questions work.

Fri Apr  6 02:11:35 IDT 2007
    Evidence domain (Ana.domains). Next is exposed questions.

Tue Apr  3 23:37:25 IDT 2007
    Active dialogue: the utility is the sum of all syndromes. It is log [P(all up)/P(all down)].
    To optimize, need domains for every evidence variable.

    Alternative hypotheses: sum of all (l-o) -- it is log [P(in this state)/P(in inverse state)].

    Foundation for these criteria? A hypothesis is that in complex systems, inverse state
    is virtually random and with low probability.

Mon Apr  2 00:56:02 IDT 2007
    Form.lhs is built on Traverse.lhs; the next step is metric computation.

Tue Mar 20 23:47:08 IST 2007
    The form generator is in the process of refactoring. Try to implement
    evidence dependencies.

Sun Jan 21 02:30:42 AMT 2007
    Anna is a good performer; next step: search by syndrome queries
    to facilitate network debugging.

Tue Jan 16 02:07:01 AMT 2007
    NetMisconf, index through ErrorT.

Sun Jan 14 01:10:41 AMT 2007
    ForeignBug drafted (even gives reasonable response).

Fri Jan 12 01:38:02 AMT 2007
    First hypothesis is breathing.

Tue Jan  9 20:46:12 AMT 2007
    Network visualization via GraphViz.

Mon Jan  8 15:41:18 AMT 2007
    syndromes in questionnaire are functional (but are not used yet).
    Next step is to manually define syndromes and hypotheses.
    
Sun Jan  7 22:19:01 AMT 2007
    entered 300 cases into ANNA-SUPPORT database.

Wed Jan  3 20:07:57 AMT 2007
    text generation for case description complete.
    Dependency on HaXml (temporary) removed.

Tue Jan  2 02:04:45 AMT 2007
    experiment in progress: ~120 cases
    text generation in Haskell drafted. 

    Next step: symbol table for questionnaire to generate
    explanation in Haskell CGI.

Fri Dec 29 02:35:11 AMT 2006
    everything is ready for the experiment.

Wed Dec 27 14:54:45 AMT 2006
    qre.lhs -- skeleton.
    Next steps: populator, javascript for collapsing and expanding
    just the questionnaire, probably without vartable yet.

Wed Dec 27 02:39:45 AMT 2006
    New modules: CGI, HTML.
    Additions to anna.synet -- just to see how it goes.
    mergeform in Sysh -- to compile standalone questionnaires.

Mon Dec 25 00:03:15 AMT 2006
    HTML part of the questionnaire is done; Regexp is removed
    from the source, no need for. Basic javascript to compose the tree.
    Now: css and js go to separate files (form.css, form.js) and an
         utility to compile html+css+js into a single document,
         collapsing/expanding (later),
         server-side cgi to gather information.

Tue Dec 19 23:57:55 AMT 2006
    HTML part of the questionnaire is mostly done.

    Next step is dynamic tree:
      Expansions are managed by JavaScript. A function traverses the tree
      and sets display:none on all disabled branches (and does not go inside).
      On enabled branches it will insert referenced sections.

Mon Dec 18 00:25:21 AMT 2006
    Html serializer skeleton.

Sat Dec 16 23:15:12 AMT 2006
    Next step is data input facility. Two things:
      1) questionnaire generator - just simple HTML.
      2) CGI script. Path info points to questionnaire,
         then interview number: quest.cgi/anna/502

         anna/ contains
            1) form.html (and other support files in the future). 
            2) numerically named data files in the form of CGI
               query strings (anna/502 is for interview #502). 

         CGI script writes updated data file when -submit-= in the query and
         returns form with data filled in.
     Will use HaXml for form-filling and a home-brewn cgi implementation.

Wed Dec 13 03:52:18 AMT 2006
    Testing framework, tests for modules.
    Shell for basic manipulations (Sysh).
    Skeleton for the questionnaire generation module.

Fri Dec  8 04:38:11 AMT 2006
    parser works on anna.synet, next is tests for the parser,
    then -- drawing the HTML.

Thu Dec  7 04:15:38 AMT 2006
    src/Parse.lhs -- parser in progress

Tue Dec  5 05:16:48 AMT 2006
    src/Synet.lhs -- basic structures and selector composition

Sun Dec  3 15:40:39 AMT 2006
    net/anna.synet drafted. exa/0.synet contains syntax patterns.

Sat Dec  2 20:38:31 AMT 2006
    Module created


src/index.js
// $Id: index.js,v 1.2 2007/01/18 07:53:01 dvd Exp $

var explanation, trprev = undefined;

function set_explanator(block) {
  var tr = block.parentNode.parentNode;
  tr.addEventListener("mouseover",
                      function(evt) {
                        if(trprev)
                          trprev.className = "case";
                        tr.className = "current case";
                        trprev = tr;
                        if(explanation.firstChild)
                          explanation.removeChild(explanation.firstChild);
                        explanation.appendChild(block.firstChild.cloneNode(false));
                      },false);
}

window.addEventListener
  ("load",
   function(evt) {
     var blocks = document.getElementsByTagName("block");
     explanation = document.getElementById("explanation");
     for(var i=0; i!=blocks.length; ++i) {
       var block = blocks[i];
       if(block.className=="explatext")
         set_explanator(block);
     }
   }, 
   false);

src/form.js
/* $Id: form.js,v 1.24 2007/05/04 09:28:29 dvd Exp $ */

/* JavaScript library used by questionnaire. */

function compatibility_hackery() {
  if(!Array.prototype.some) {
    Array.prototype.some =  function(f) {
      for(var i=0; i!=this.length; ++i) {
        if(f(this[i])) return true;
      }
      return false;
    }
  }
  
  if(!Array.prototype.map) {
    Array.prototype.map = function(f) {
      var a = new Array();
      for(var i=0; i!=this.length; ++i) {
        a.push(f(this[i]));
      }
      return a;
    }
  }
}
compatibility_hackery();


function tag_name(node) {
  return node.tagName.toLowerCase();
}

function text_content(node) {
  if(node.textContent!=undefined)
    return node.textContent;
  else
    return node.childNodes[0].nodeValue;
}

function attach_uncheckers_to_radios() {
  /* when the user clicks on a checked radio, uncheck it */
  var inputs = document.getElementsByTagName("input");

  for(var i=0;i!=inputs.length; ++i) {
    var node = inputs[i];
    if(node.attributes["type"].value=="radio") {
      node.addEventListener
        ("mousedown",
         function(evt) {
           this.prev_checked = this.checked;
         },
         false);
      node.addEventListener
        ("click",
         function(evt) {
           this.checked = !this.prev_checked;
         },
         false);
    }
  }
}

function bind_checkboxes () {
  /* for every checkbox, link its onchange handler to the paired
     hidden field */
  var inputs = document.getElementsByTagName("input"),
      hidtab = {};

  function linkhidden(hidden) {
    return function(evt) {
      switch(hidden.value) {
      case "0":
      case "1": hidden.value = "2"; break;
      case "2": hidden.value = "1"; break;
      }
    }
  }

  for(var i=0; i!=inputs.length; ++i) {
    if(inputs[i].attributes["type"].value=="hidden")
      hidtab[inputs[i].name] = inputs[i];
  }

  for( var i=0; i!=inputs.length; ++i) {
    if(inputs[i].attributes["type"].value=="checkbox") {
      inputs[i].addEventListener("click",
                                 linkhidden(hidtab[inputs[i].name]),false);
    }
  }
}

// evidence & syndrome value retrieval 
var memotab = {};

function memoize(f) {
  /* a dictionary of memoized syndrome values is attached
     to the current window, every value is computed once
     per pass */
  return function(x) {
    if(x in memotab) return memotab[x];
    else return memotab[x] = f(x);
  }
}
 
var value = memoize
  (function(name) {
    /* retrieve variable value from the document. */
    var syval = syndrome(name,value); /* syndrome computation is
                                         defined elsewhere */
    if(syval!=undefined) return syval;
    var byname = document.getElementsByName(name);
    
    for(var i=0; i!=byname.length; ++i) {
      var node = byname[i];
      if(tag_name(node)=="input") {
        switch(node.attributes["type"].value) {
        case "text": return node.value==""?0:1;
        case "hidden": return parseFloat(node.value);
        case "radio":
          if(node.checked)
            return parseFloat(node.value);
          else
            break;
        }
      }
    }

    return 0;
  });

syndrome = function (name,value) {
  if(name in syndromes) return syndromes[name](value);
  return undefined;
}

function reshape_subtree(node) {
  /* for every node, get its children. for
     div[@class=section-reference] insert section.
     for div[@class=expansion] traverse children and expand
     those that qualify. For the rest, just call
     reshape_quesionnaire(child). */

  function selected(val,div) {
    /* tests whether the division is selected */
    var guards = new Array();
    for(var i=0;i!=div.childNodes.length; ++i) {
      var node = div.childNodes[i];
      if(node.nodeType==1
         && tag_name(node)=="span"
         && node.className=="guard") {
        guards.push(node);
      }
    }

    return guards
      .some(function(guard) {
          var exp = text_content(guard),
            op = exp.substring(0,2),
            arg = exp.substring(2);
          switch(op) {
          case "**": return true;
          case "eq": return val==parseFloat(arg);
          case "le": return val<=parseFloat(arg);
          case "ge": return val>=parseFloat(arg);
          case "in":
            var bounds = arg.split(' ').map(parseFloat);
            return bounds[0]<=val && val <= bounds[1];
          }
        });
  }

  if(node.nodeType==1) {
    if(tag_name(node)=="div" && node.className=="section-reference") {
      var sectId;
      for(var i = 0; i!=node.childNodes.length; ++i) {
        var child = node.childNodes[i];
        if(child.nodeType==1
           && tag_name(child)=="span"
           && child.className=="ref-id") {
          sectId = child.childNodes[0].nodeValue;
          break;
        }
      }
      var section = document.getElementById(sectId);
      node.appendChild(section);
      reshape_subtree(section);
    } else if(tag_name(node)=="div" && node.className=="expansion") {
      var val;
      for(var i=0; i!= node.childNodes.length; ++i) {
        var child = node.childNodes[i];
        if(child.nodeType==1) {
          if(tag_name(child)=="span"
             && child.className=="ref-name") {
            val = value(text_content(child));
          } else if(tag_name(child)=="div"
                    && child.className=="selection") {
            if(selected(val,child)) {
              child.style.display = "block";
              reshape_subtree(child);
            } else {
              child.style.display = "none";
            }
          }
        }
      }
    } else {
      for(var i = 0; i!=node.childNodes.length; ++i) {
        reshape_subtree(node.childNodes[i]);
      }
    }
  }
}

function mark_and_sweep (node) {
  /* reset hidden input elements to defaults */
  var active_inputs = {}, s = "", t = "";
  
  function mark(node) {
    if(node.nodeType==1) {
      if(node.style.display!="none") {
        if(tag_name(node)=="input") {
          active_inputs[node.name] = true;
        }
        for(var i=0;i!=node.childNodes.length; ++i) {
          mark(node.childNodes[i]);
        }
      }
    }
  }

  function reset(input) {
    /* reset input element to default (unanswered) state; must be
       in agreement with Form.lhs:instance Ht Context Answer */
    switch(input.attributes["type"].value) {
    case "text": input.value = ""; break;
    case "hidden": if(input.className=="hidcheck") input.value = "0"; break;
    case "radio": case "checkbox": input.checked = false; break;
    }
  }

  mark(node);
  
  var inputs = document.getElementsByTagName("input");
  for(var i=0;i!=inputs.length;++i) {
    s+= inputs[i].name+" ";
    if(!(inputs[i].name in active_inputs)) {
      t+= inputs[i].name+" ";
      reset(inputs[i]);
    }
  }
}

function reshape_questionnaire() {
  memotab = {};
  var soil = document.getElementById("se--soil");
  reshape_subtree(soil);
  mark_and_sweep(soil);
}

function reshape_on_input() {
  var inputs = document.getElementsByTagName("input");
  for(var i=0;i!=inputs.length; ++i) {
    node = inputs[i];
    var event;
    switch(node.attributes["type"].value) {
    case "radio":
    case "checkbox": event = "click"; break;
    case "text": event = "change"; break;
    }
    node.addEventListener(event,reshape_questionnaire,false);
  }
}

function activate_alternative_explanations() {
  var expladiv = document.getElementById("explanation"),
      as = document.getElementsByTagName("a"),
      onclick =
        function (evt) {
          if(expladiv.firstChild) expladiv.removeChild(expladiv.firstChild);
          expladiv.appendChild(this.getElementsByTagName("block")[0].firstChild.cloneNode(false));
        };
  for(var i=0;i!=as.length;++i) {
    node = as[i];
    if(node.className=='explalink') 
      node.addEventListener("click",onclick,false);
  }
}

function hide_compatibility_banner() {
  var banner = document.getElementById("loading-banner");
  banner.parentNode.removeChild(banner);
}

window.addEventListener
  ("load",
   function (evt) {
     attach_uncheckers_to_radios(document.body);
     bind_checkboxes(document.body);
     reshape_questionnaire();
     reshape_on_input();
     activate_alternative_explanations();
     hide_compatibility_banner();
   },
   false);
